src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1447,26 +1461,115 @@ protected override void AfterLeftChildHasBeenVisited(BoundBinaryOperator binary)
 
                         if (operandComparedToNull.Type?.IsReferenceType == true)
                         {
-                            int slot = MakeSlot(operandComparedToNull);
+                            ImmutableArray<int> slots = getOperandSlots(operandComparedToNull);
+                            if (!slots.IsEmpty)
+                            {
+                                Split();
+                                foreach (int slot in slots)
+                                {
+                                    if (op == BinaryOperatorKind.Equal)
+                                    {
+                                        this.StateWhenFalse[slot] = true;
+                                    }
+                                    else
+                                    {
+                                        this.StateWhenTrue[slot] = true;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            ImmutableArray<int> getOperandSlots(BoundExpression operand)
```

Comment:
Consider having the caller pass an `ArrayBuilder` argument rather than creating an `ImmutableArray` result.